vmdb/spec/models/metric/ci_mixin/capture/openstack_infra_spec.rb

```diff
@@ -46,56 +45,911 @@
+      @ts_keys = @values_by_ts.keys.sort
+    end
+
+    ###################################################################################################################
+    # DESCRIPTION FOR: net_usage_rate_average
+    # MAIN SCENARIOS :
+    # "hardware.network.ip.incoming.datagrams", "hardware.network.ip.incoming.datagrams" are two streams of data whose
+    # values should be summed together and divided by intervals between their last values. We should observe in these
+    # tests, that incomplete data buckets (data from all streams) from borders of collection periods should be thrown
+    # away and collected in next period. Incomplete data buckets from middle of the stream should be replaced by next
+    # complete bucket.
+    #
+    # 1. The streams has different sizes, that means Ceilometer was not able to collect data for both streams at some
+    # point. We should verify that these incomplete values (we have value only from one stream) are thrown away, and
+    # the whole period should be filled with next complete value (that has value from both streams). Also warning that
+    # data are corrupted should be raised.
+    #
+    # ASCII DIAGRAM: streams a and b, value for for 4b is missing, 4a should be thrown away and replaced by 5, which is
+    # complete. The whole (3, 5> interval will be filled with value of 5.
+    # NOTE: diagram is just for explaining of the process, real data are different
+    #
+    # stream a:         ---------1a-----2a-------------3a----------4a--------5a
+    # stream b:         ------1b----------2b--------3b--------------------5b
+    # saved stream:           1---------2-----------3------------------------5
+    #
+    # 2. We will observe collecting of the the data in 2 collection periods. When in first period, the last stats value
+    # will have value from only one stream, so it should be thrown away. Since it's in the end of the collection period,
+    # the stats will not be filled in that period.
+    #
+    # ASCII DIAGRAM: streams a and b, 1. collection period is missing sample 3a, so the whole (2, 3> period should not
+    # be saved. Then 2. collection period start is moved default 20 minutes back, so this period is able to save the
+    # missing (2, 3> period. We need to be able to obtain sample number 2, for computing last_value for diff.
+    # NOTE: diagram is just for explaining of the process, real data are different
+    #
+    # stream a 1.collection period:    --------------1a-----2a----------------
+    # stream b 1.collection period:    ----------1b-----------2b----------3b--
+    # stream a 2.collection period:                  1a-----2a----------------3a-------4a--------5a-
+    # stream b 2.collection period:                  ---------2b----------3b-------------------5b---
+    # saved stream 1.collection period:          1----------2
+    # saved stream 2.collection period:                      -------------3--------------------5
```

Comment:
I think I'd like to see these ascii diagrams replicated in both the mainline code above (probably as rdoc in `Metric::CiMixin::Capture::OpenstackBase`) and also in guides.

I know that we want to drop the 20-sec alignment thing at some point, but having this level of detailed description is huge in being able to explain what the heck is going on with all this code.

Additionally, if we're ever asked why it's such a huge deal to drop the 20-sec requirement, we can point to these documents to show just how complicated the logic is and what a maintenance nightmare this is.